README.md

[![Arduino CI](https://github.com/RobTillaart/AD568X/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/AD568X/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/AD568X/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/AD568X/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/AD568X/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/AD568X.svg)](https://github.com/RobTillaart/AD568X/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/AD568X/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/AD568X.svg?maxAge=3600)](https://github.com/RobTillaart/AD568X/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/AD568X.svg)](https://registry.platformio.org/libraries/robtillaart/AD568X)


# AD568X

Arduino library for the AD568X series digital analog convertor.


## Description

**Experimental**

- SPI interface  (AD569X series is I2C)

Note: not yet tested, 
TODO buy hardware.
TODO documentation.


|  Type     |  bits  |  supported  |  Notes  |
|:---------:|:------:|:-----------:|:--------|
|  AD5680   |   18   |     N       |  has separate class
|  AD5681R  |   12   |     Y       |
|  AD5682R  |   14   |     Y       |
|  AD5683   |   16   |     Y       |
|  AD5683R  |   16   |     Y       |


Feedback, issues, improvements are welcome. 
Please file an issue on GitHub.


## Related

- https://github.com/RobTillaart/AD56x8 (multi channel)
- https://github.com/RobTillaart/AD5680 (18 bit version)
- https://github.com/RobTillaart/MCP_DAC (SPI interface)


## Interface

```cpp
#include "AD568X.h"
```

### Base class

Should not be used to instantiate a device as the derived types have 
set the correct number of bits. 

- **AD568X(uint8_t slaveSelect)** constructor base class, sets HW SPI.
Sets internal values to zero.
- **AD568X(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect)** constructor, 
sets SW SPI.
Sets internal values to zero.
- **begin()** initializes the SPI and sets internal state.


#### Derived classes (preferred use)

The parameters for the specific constructors are identical to the base class.
One should use these!

- **AD5681R(..)** constructor, 12 bit.
- **AD5682R(..)** constructor, 14 bit.
- **AD5683R(..)** constructor, 16 bit.
- **AD5683(..)** constructor,  16 bit.


### Set DAC

- **bool setValue(uint16_t value)** set value to the output immediately, 
effectively a prepare + update in one call.
Returns false if value out of range.
- **uint16_t getValue()** returns set OR prepared value.
At power up the DAC's will be reset to 0 Volt.
- **bool setPercentage(float percentage)** idem.
- **float getPercentage()** idem.
- **bool prepareValue(uint16_t value)** prepares the value.
Returns false if value out of range.
- **bool updateValue()** writes the prepared value to ADC.
Returns false if value out of range.


### Control Register

Check datasheet for details.

- **bool setControlRegister(uint16_t value)** set register in one call.
- **uint16_t getControlRegister()** return the set / current value.
- **bool reset()** reset the AD568X device.
Also resets all flags in control register.
- **bool setPowerDownMode(uint8_t mode = AD568X_PWR_NORMAL)** 
set one of three power down modes, see table below, or to normal mode.
Default is AD568X_PWR_NORMAL.
- **bool disableReference(bool b)** false = reference enabled.
true = reference disabled (uses less power). 
Default is enabled.
- **bool enableGain(bool enable = false)** false = 0..Vref  true = 0..2x Vref.
Default is false.
- **bool enableDaisyChain(bool enable = false)** enables daisy chain mode.
**WARNING** Daisy Chain is not supported in current library.
Default is false.

Control register bits: (see datasheet)

|  bits  |  meaning             |  Notes  |
|:------:|:---------------------|:--------|
|   15   |  RESET               |
|   14   |  PowerDown 1         |
|   13   |  PowerDown 2         |
|   12   |  REFerence selection |
|   11   |  GAIN                |
|   10   |  DCEN, Daisy Chain   |  not supported in the library.


Power down modi: (see datasheet)

|  mode  |  define             |
|:------:|:--------------------|
|  0x00  |  AD568X_PWR_NORMAL  |
|  0x01  |  AD568X_PWR_1K      |
|  0x02  |  AD568X_PWR_100K    |
|  0x03  |  AD568X_PWR_TRI     |




#### SPI 

- **void setSPIspeed(uint32_t speed)** sets SPI clock in **Hz**,
please read datasheet of the ADC first to get optimal speed.
- **uint32_t getSPIspeed()** gets current speed in **Hz**.
- **bool usesHWSPI()** returns true if HW SPI is used.


#### SPI ESP32 specific

("inherited" from MPC_DAC library)

- **void selectHSPI()** in case hardware SPI, the ESP32 has two options HSPI and VSPI.
- **void selectVSPI()** see above.
- **bool usesHSPI()** returns true if HSPI is used.
- **bool usesVSPI()** returns true if VSPI is used.

The **selectVSPI()** or the **selectHSPI()** needs to be called 
BEFORE the **begin()** function.

(experimental)
- **void setGPIOpins(uint8_t clk, uint8_t miso, uint8_t mosi, uint8_t select)** 
overrule GPIO pins of ESP32 for hardware SPI. Needs to be called AFTER the **begin()** function.

Note: earlier experiments shows that on a ESP32 SW-SPI is equally fast as 
HW-SPI and in fact a bit more stable. 
The SW pulses are a bit slower than the HW pulses and therefore more square. 
The HW-SPI has some overhead SW-SPI hasn't. 


## Future

#### Must

- improve documentation
- get test hardware
- test the library
- support for LDAC pin
  - void **setLDACPin(uint8_t pin)**
  - void **triggerLDAC()**


#### Should

- write examples
  - performance


#### Could

- write unit test


#### Wont


## Support

If you appreciate my libraries, you can support the development and maintenance.
Improve the quality of the libraries by providing issues and Pull Requests, or
donate through PayPal or GitHub sponsors.

Thank you,





CHANGELOG.md
# Change Log AD56X8

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.0] - 2023-09-18
- initial work





test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2023-09-18
// PURPOSE: unit tests for the AD568X DA convertor.
//          https://github.com/RobTillaart/AD568X
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)


#include <ArduinoUnitTests.h>

#include "Arduino.h"
#include "AD568X.h"



unittest_setup()
{
  fprintf(stderr, "AD568X_LIB_VERSION: %s\n", (char *) AD568X_LIB_VERSION);
}

unittest_teardown()
{
}


unittest(constant)
{
  assertEqual(0x00, AD568X_PWR_NORMAL    );
  assertEqual(0x01, AD568X_PWR_1K        );
  assertEqual(0x02, AD568X_PWR_100K      );
  assertEqual(0x03, AD568X_PWR_TRI_STATE );
}


unittest(constructors)
{
  AD5681R AD0(8);           //  12 bit  0..4096
  AD5681R AD1(8, 9, 10);    //  12 bit  0..4096

  assertTrue(AD0.usesHWSPI());
  assertFalse(AD1.usesHWSPI());
}


unittest(get_type)
{
  AD568X  ADX(8);
  AD5681R AD1R(8);
  AD5682R AD2R(8);
  AD5683  AD3(8);
  AD5683R AD3R(8);

  assertEqual(00, ADX.getType());
  assertEqual(12, AD1R.getType());
  assertEqual(14, AD2R.getType());
  assertEqual(16, AD3.getType());
  assertEqual(16, AD3R.getType());
}


unittest(get_setValue)
{
  AD5681R AD1R(8);

  AD1R.begin();
  for (int v = 0; v < 2000; v += 100)
  {
    AD1R.setValue(v);
    assertEqual(v, AD1R.getValue());
  }
}


unittest(get_setPercentage)
{
  AD5681R AD1R(8);

  AD1R.begin();
  for (int p = 0; p < 100; p += 9)
  {
    AD1R.setPercentage(p);
    assertEqual(p, AD1R.getPercentage());
  }
}


unittest_main()


//  -- END OF FILE --


AD568X.cpp
//
//    FILE: AD568X.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.0
//    DATE: 2023-09-18
// PURPOSE: Arduino library for AD568X series Digital Analog Convertor.


#include "AD568X.h"

//  not all "commands" implemented

#define AD568X_REG_NOP            0x00
#define AD568X_REG_WRITE          0x10
#define AD568X_REG_UPDATE         0x20
#define AD568X_REG_WRITE_UPDATE   0x30
#define AD568X_REG_CONTROL        0x40


AD568X::AD568X(uint8_t slaveSelect)
{
  _hwSPI  = true;
  _select = slaveSelect;
  _value  = 0;
}


AD568X::AD568X(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect)
{
  _hwSPI   = false;
  _dataOut = spiData;
  _clock   = spiClock;
  _select  = slaveSelect;
  _value   = 0;
}


//  initializes the SPI
//  and sets internal state
void AD568X::begin()
{
  pinMode(_select, OUTPUT);
  digitalWrite(_select, HIGH);

  _spi_settings = SPISettings(_SPIspeed, MSBFIRST, SPI_MODE1);

  if(_hwSPI)
  {
    #if defined(ESP32)
    if (_useHSPI)      //  HSPI
    {
      mySPI = new SPIClass(HSPI);
      mySPI->end();
      mySPI->begin(14, 12, 13, _select);   //  CLK=14  MISO=12  MOSI=13
    }
    else               //  VSPI
    {
      mySPI = new SPIClass(VSPI);
      mySPI->end();
      mySPI->begin(18, 19, 23, _select);   //  CLK=18  MISO=19  MOSI=23
    }
    #else              //  generic hardware SPI
    mySPI = &SPI;
    mySPI->end();
    mySPI->begin();
    #endif
    delay(1);
  }
  else                 //  software SPI
  {
    pinMode(_dataOut, OUTPUT);
    pinMode(_clock, OUTPUT);
    digitalWrite(_dataOut, LOW);
    digitalWrite(_clock, LOW);
  }
}


uint8_t AD568X::getType()
{
  return _type;
}


void AD568X::setLDACPin(uint8_t ldac)
{
  _ldac = ldac;
  pinMode(ldac, OUTPUT);
  digitalWrite(ldac, HIGH);
}


bool AD568X::triggerLDAC()
{
  if (_ldac == 255) return false;
  digitalWrite(ldac, LOW);
  digitalWrite(ldac, HIGH);
}


//  value = 0..65535 (16 bit), 16383 (14 bit), 4095 (12 bit) depending on type)
bool AD568X::setValue(uint16_t value)
{
  if ((_type == 12) && (value > 4095))  return false;
  if ((_type == 14) && (value > 16383)) return false;
  _value = value;
  updateDevice(AD568X_REG_WRITE_UPDATE, value);
  return true;
}


//  returns 0..65535 (16 bit), 16383 (14 bit), 4095 (12 bit) depending on type)
uint16_t AD568X::getValue()
{
  return _value;
}


bool AD568X::setPercentage(float percentage)
{
  uint16_t value = 0;
  if ((percentage < 0) || (percentage > 100)) return false;
  if      (_type == 16) value = round(655.35 * percentage);
  else if (_type == 14) value = round(163.83 * percentage);
  else /* type = 12 */  value = round( 40.95 * percentage);
  return setValue(value);
}


float AD568X::getPercentage()
{
  float value = getValue();
  if (value > 0)
  {
    if (_type == 16) return value * ( 1.0 / 655.35);
    if (_type == 14) return value * ( 1.0 / 163.83);
    if (_type == 12) return value * ( 1.0 /  40.95);
  }
  return 0;
}


bool AD568X::prepareValue(uint16_t value)
{
  if ((_type == 12) && (value > 4095))  return false;
  if ((_type == 14) && (value > 16383)) return false;
  _value = value;
  updateDevice(AD568X_REG_WRITE, value);
  return true;
}


bool AD568X::updateValue()
{
  updateDevice(AD568X_REG_UPDATE, 0);
  return true;
}


//////////////////////////////////////////////////////////////////
//
//  CONTROL REGISTER
//
bool AD568X::setControlRegister(uint16_t value)
{
  _controlReg = value & 0xFC00;
  updateDevice(AD568X_REG_CONTROL, _controlReg);
  return true;
}


bool AD568X::reset()
{
  _controlReg |= (1 << 15);  //  set RESET bit
  updateDevice(AD568X_REG_CONTROL, _controlReg);
  _controlReg = 0x0000;
  return true;
}


bool AD568X::setPowerDownMode(uint8_t mode)
{
  if (mode > 3) return false;
  _controlReg &= 0x9C00;  //  clear PD bits
  _controlReg |= (mode << 13);
  updateDevice(AD568X_REG_CONTROL, _controlReg);
  return true;
}


bool AD568X::disableReference(bool b)
{
  _controlReg &= 0xEC00;  //  clear REF bit
  if (b) _controlReg |= (1 << 12);
  updateDevice(AD568X_REG_CONTROL, _controlReg);
  return true;
}


bool AD568X::enableGain(bool enable)
{
  _controlReg &= 0xF400;  //  clear GAIN bit
  if (enable) _controlReg |= (1 << 11);
  updateDevice(AD568X_REG_CONTROL, _controlReg);
  return true;
}


bool AD568X::enableDaisyChain(bool enable)
{
  _controlReg &= 0xF800;  //  clear DCEN bit
  if (enable) _controlReg |= (1 << 10);
  updateDevice(AD568X_REG_CONTROL, _controlReg);
  return true;
}


//////////////////////////////////////////////////////////////////
//
//  SPI
//
void AD568X::setSPIspeed(uint32_t speed)
{
  _SPIspeed = speed;
  _spi_settings = SPISettings(_SPIspeed, MSBFIRST, SPI_MODE1);
};


uint32_t AD568X::getSPIspeed()
{
  return _SPIspeed;
};


bool AD568X::usesHWSPI()
{
  return _hwSPI;
}


//  ESP32 specific
#if defined(ESP32)

void AD568X::selectHSPI()
{
  _useHSPI = true;
}


void AD568X::selectVSPI()
{
  _useHSPI = false;
}


bool AD568X::usesHSPI()
{
  return _useHSPI;
}


bool AD568X::usesVSPI()
{
  return !_useHSPI;
}


void AD568X::setGPIOpins(uint8_t clk, uint8_t miso, uint8_t mosi, uint8_t select)
{
  _clock   = clk;
  _dataOut = mosi;
  _select  = select;
  pinMode(_select, OUTPUT);
  digitalWrite(_select, HIGH);

  mySPI->end();                            //  disable SPI
  mySPI->begin(clk, miso, mosi, select);   //  enable SPI
}

#endif


//////////////////////////////////////////////////////////////////
//
//  PRIVATE
//

void AD568X::updateDevice(uint8_t cmd, uint16_t value)
{
  //  to be optimized - switch cmd etc
  uint8_t a = cmd | (value >> 12);
  uint8_t b = (value >> 4) & 0xFF;
  uint8_t c = (value << 4) & 0xF0;

  updateDevice(a, b, c);
}


void AD568X::updateDevice(uint8_t a, uint8_t b, uint8_t c)
{
  digitalWrite(_select, LOW);
  if (_hwSPI)
  {
    mySPI->beginTransaction(_spi_settings);
    mySPI->transfer(a);
    mySPI->transfer(b);
    mySPI->transfer(c);
    mySPI->endTransaction();
  }
  else //  Software SPI
  {
    swSPI_transfer(a);
    swSPI_transfer(b);
    swSPI_transfer(c);
  }
  digitalWrite(_select, HIGH);
}


//  simple one mode version
void AD568X::swSPI_transfer(uint8_t value)
{
  uint8_t clk = _clock;
  uint8_t dao = _dataOut;
  for (uint8_t mask = 0x80; mask; mask >>= 1)
  {
    digitalWrite(dao,(value & mask));
    digitalWrite(clk, HIGH);
    digitalWrite(clk, LOW);
  }
}


/////////////////////////////////////////////////////////////////////////////
//
//  DERIVED  AD5680
//
// AD5680::AD5680(uint8_t slaveSelect) : AD568X(slaveSelect)
// {
  // _type  = 18;
  // _value = 0;
// }

// AD5680::AD5680(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect)
                // : AD568X(spiData, spiClock, slaveSelect)
// {
  // _type  = 18;
  // _value = 0;
// }


/////////////////////////////////////////////////////////////////////////////
//
//  DERIVED  AD5681
//
AD5681R::AD5681R(uint8_t slaveSelect) : AD568X(slaveSelect)
{
  _type  = 12;
  _value = 0;
}

AD5681R::AD5681R(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect)
                : AD568X(spiData, spiClock, slaveSelect)
{
  _type = 12;
  _value = 0;
}


/////////////////////////////////////////////////////////////////////////////
//
//  DERIVED  AD5682
//
AD5682R::AD5682R(uint8_t slaveSelect) : AD568X(slaveSelect)
{
  _type  = 14;
  _value = 0;
}

AD5682R::AD5682R(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect)
                : AD568X(spiData, spiClock, slaveSelect)
{
  _type  = 14;
  _value = 0;
}


/////////////////////////////////////////////////////////////////////////////
//
//  DERIVED  AD5683
//
AD5683R::AD5683R(uint8_t slaveSelect) : AD568X(slaveSelect)
{
  _type  = 16;
  _value = 0;
}

AD5683R::AD5683R(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect)
                : AD568X(spiData, spiClock, slaveSelect)
{
  _type  = 16;
  _value = 0;
}


/////////////////////////////////////////////////////////////////////////////
//
//  DERIVED  AD5683
//
AD5683::AD5683(uint8_t slaveSelect) : AD568X(slaveSelect)
{
  _type  = 16;
  _value = 0;
}

AD5683::AD5683(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect)
                : AD568X(spiData, spiClock, slaveSelect)
{
  _type  = 16;
  _value = 0;
}


//  -- END OF FILE --



AD568X.h
#pragma once
//
//    FILE: AD568X.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.0
//    DATE: 2023-09-18
// PURPOSE: Arduino library for AD568X series Digital Analog Convertor.


#include "Arduino.h"
#include "SPI.h"

#define AD568X_LIB_VERSION        (F("0.1.0"))


#define AD568X_PWR_NORMAL         0x00
#define AD568X_PWR_1K             0x01
#define AD568X_PWR_100K           0x02
#define AD568X_PWR_TRI_STATE      0x03


class AD568X
{
public:
  AD568X(uint8_t slaveSelect);
  AD568X(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect);

  void     begin();
  uint8_t  getType();


  //  SET DAC
  //  sets output immediately = prepare + update.
  bool     setValue(uint16_t value);
  uint16_t getValue();
  bool     setPercentage(float percentage);
  float    getPercentage();


  //  prep & update separately
  bool     prepareValue(uint16_t value);
  bool     updateValue();


  //  CONTROL REGISTER (see datasheet)
  //  bit     meaning
  //   15     reset
  //   14     power down 1
  //   13     power down 0
  //   12     REFerence selection
  //   11     GAIN
  //   10     DCEN, Daisy Chain ENable
  //  other   0
  //  set register in one call.
  bool     setControlRegister(uint16_t value);
  uint16_t getControlRegister();

  bool     reset();
  bool     setPowerDownMode(uint8_t mode = AD568X_PWR_NORMAL);
  bool     disableReference(bool b);
  bool     enableGain(bool enable = false);
  bool     enableDaisyChain(bool enable = false);


  //  SPI
  //       speed in Hz
  void     setSPIspeed(uint32_t speed);
  uint32_t getSPIspeed();
  bool     usesHWSPI();

  //  ESP32 specific
  #if defined(ESP32)
  void     selectHSPI();
  void     selectVSPI();
  bool     usesHSPI();
  bool     usesVSPI();

  //  to overrule ESP32 default hardware pins
  void     setGPIOpins(uint8_t clk, uint8_t miso, uint8_t mosi, uint8_t select);
  #endif


protected:
  uint8_t  _type        = 0;      //  # bits
  uint8_t  _dataOut     = 255;
  uint8_t  _clock       = 255;
  uint8_t  _select      = 255;

  uint16_t _value       = 0;
  uint16_t _controlReg  = 0;

  bool     _hwSPI       = false;
  uint32_t _SPIspeed    = 16000000;

  void     updateDevice(uint8_t cmd, uint16_t value);
  void     updateDevice(uint8_t a, uint8_t b, uint8_t c);
  void     swSPI_transfer(uint8_t value);

  SPIClass    * mySPI;
  SPISettings _spi_settings;

  #if defined(ESP32)
  bool        _useHSPI = true;
  #endif
};


/////////////////////////////////////////////////////////////////////////////
//
//  DERIVED CLASSES
//
// class AD5680 : public AD568X   need 32 bit value...etc
// {
// public:
  // AD5680(uint8_t slaveSelect);
  // AD5680(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect);
// };


class AD5681R : public AD568X
{
public:
  AD5681R(uint8_t slaveSelect);
  AD5681R(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect);

  //  SHORT WRITE ?
};


class AD5682R : public AD568X
{
public:
  AD5682R(uint8_t slaveSelect);
  AD5682R(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect);
};


class AD5683R : public AD568X
{
public:
  AD5683R(uint8_t slaveSelect);
  AD5683R(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect);
};


class AD5683 : public AD568X
{
public:
  AD5683(uint8_t slaveSelect);
  AD5683(uint8_t spiData, uint8_t spiClock, uint8_t slaveSelect);
};



// -- END OF FILE --
